### PR TITLE
Since log request and delivery request client_request_ids\TESTING=unit

### PR DESCRIPTION
**Pull request by scottmcmaster**

### CODE REVIEW

File: lib/promoted/ruby/client.rb [hunk 1]
```diff
@@ -168,7 +168,8 @@ def deliver args, headers={}
             log_request = {
               :full_insertion => response_insertions,
               :experiment => cohort_membership_to_log,
-              :request => request_to_log
+              :request => request_to_log,
+              :client_request_id => delivery_request_builder.client_request_id
```
Comment (danbosnichill) [inline lib/promoted/ruby/client.rb:172]: client_request_id is a field on Request, not LogRequest.  Ditto down below.

File: lib/promoted/ruby/client/request_builder.rb [hunk 2]
```diff
@@ -96,7 +102,8 @@ def log_request_params(include_insertions: true, include_request: true)
             user_info: user_info,
             timing: timing,
             cohort_membership: @experiment,
-            client_info: @client_info
+            client_info: @client_info,
+            client_request_id: request[:client_request_id] || @client_request_id # if logging a request, match the client request id, else use the one we generated
```
Comment (danbosnichill) [inline lib/promoted/ruby/client/request_builder.rb:106]: ditto

**Review by danbosnichill — CHANGES_REQUESTED** [review]

File: lib/promoted/ruby/client.rb (hunk 1 above)
Comment (scottmcmaster) [reply lib/promoted/ruby/client.rb:172]: Generally speaking, this is one of the more challenging parts of programming against LogRequest especially, namely, what is going on at the top level vs. what is going on in the nested "request" field.

I think there might also be some misunderstanding of how the Ruby code works. The structure of the final messages that get send to Metrics API and Delivery API respectively are built in RequestBuilder.log_request_params and RequestBuilder.delivery_request_params. Hop over there and check it out, but, in each case, client_request_id ends up at the top level. (You can also see this in the diff of request_builder_spec.rb.)

Can you reconfirm that this is not correct, and I guess in that case we expect client_request_id to be on the _inner_ request (that only gets filled in when !insertions_from_delivery. Thanks.

Comment (scottmcmaster) [reply lib/promoted/ruby/client.rb]: Updated per our conversation, please look again (especially the test cases that show the usage). Thanks!

File: lib/promoted/ruby/client/request_builder.rb [hunk 3]
```diff
@@ -99,10 +103,11 @@ def log_request_params(include_insertions: true, include_request: true)
             client_info: @client_info
           }
 
-          request[:request_id] = @id_generator.newID if not request[:request_id]
-
           # Log request allows for multiple requests but here we only send one.
-          params[:request] = [request] if include_request
+          if include_request
+            request[:request_id] = request[:request_id] || @id_generator.newID
```
Comment (danbosnichill) [inline lib/promoted/ruby/client/request_builder.rb:108]: It's hard for me to tell with ruby client if this covers all of the cases that I'd expect

**Review by danbosnichill — APPROVED** [review]

File: lib/promoted/ruby/client/request_builder.rb (hunk 3 above)
Comment (scottmcmaster) [reply lib/promoted/ruby/client/request_builder.rb:108]: ftr in practice, this is the "SDK-did-delivery" case.

**Pull request merged**